crates/ruff_server/src/server/api/notifications/did_close.rs

```diff
@@ -22,20 +22,24 @@ impl super::SyncNotificationHandler for DidClose {
             text_document: types::TextDocumentIdentifier { uri },
         }: types::DidCloseTextDocumentParams,
     ) -> Result<()> {
-        // Publish an empty diagnostic report for the document if the client does not support pull diagnostics.
-        // This will de-register any existing diagnostics.
-        if !session.resolved_client_capabilities().pull_diagnostics {
-            let snapshot = session
-                .take_snapshot(&uri)
-                .ok_or_else(|| {
-                    anyhow::anyhow!("Unable to take snapshot for document with URL {uri}")
-                })
-                .with_failure_code(lsp_server::ErrorCode::InternalError)?;
-            clear_diagnostics_for_document(&snapshot, &notifier)?;
+        // Publish an empty diagnostic report for the document. This will de-register any existing diagnostics.
+        let snapshot = session
+            .take_snapshot(&uri)
+            .ok_or_else(|| anyhow::anyhow!("Unable to take snapshot for document with URL {uri}"))
+            .with_failure_code(lsp_server::ErrorCode::InternalError)?;
+        clear_diagnostics_for_document(snapshot.query(), &notifier)?;
+
+        let key = snapshot.query().make_key();
+
+        // Notebook cells will go through the `textDocument/didClose` path.
+        // We still want to publish empty diagnostics for them, but we
+        // shouldn't call `session.close_document` on them.
+        if matches!(key, DocumentKey::NotebookCell(_)) {
+            return Ok(());
         }
```

Comment:
Huh, I wasn't aware of this but it actually does send the requests to close each cell first and then the notebook:
```
2024-05-15 19:19:48.235 [info] [Trace - 7:19:48 PM] Sending notification 'textDocument/didClose'.
2024-05-15 19:19:48.236 [info] [Trace - 7:19:48 PM] Sending notification 'textDocument/didClose'.
2024-05-15 19:19:48.236 [info] [Trace - 7:19:48 PM] Sending notification 'textDocument/didClose'.
2024-05-15 19:19:48.236 [info] [Trace - 7:19:48 PM] Sending notification 'textDocument/didClose'.
2024-05-15 19:19:48.236 [info] [Trace - 7:19:48 PM] Sending notification 'textDocument/didClose'.
2024-05-15 19:19:48.236 [info] [Trace - 7:19:48 PM] Sending notification 'notebookDocument/didClose'.
```

I guess it's not a problem in `ruff-lsp` because of the way the Notebook and cell documents are modeled in `pygls`.